chain/store/splitstore/splitstore.go

```diff
@@ -0,0 +1,422 @@
+
+	default:
+		return nil, err
+	}
+}
+
+func (s *SplitStore) GetSize(cid cid.Cid) (int, error) {
+	size, err := s.hot.GetSize(cid)
+
+	switch err {
+	case nil:
+		return size, nil
+
+	case bstore.ErrNotFound:
+		return s.cold.GetSize(cid)
+
+	default:
+		return 0, err
+	}
+}
+
+func (s *SplitStore) Put(blk blocks.Block) error {
+	s.mx.Lock()
+	epoch := s.curTs.Height()
+	s.mx.Unlock()
+
+	err := s.snoop.Put(blk.Cid(), epoch)
+	if err != nil {
+		log.Errorf("error tracking CID in hotstore: %s; falling back to coldstore", err)
+		return s.cold.Put(blk)
+	}
+
+	return s.hot.Put(blk)
+}
+
+func (s *SplitStore) PutMany(blks []blocks.Block) error {
+	err := s.hot.PutMany(blks)
+	if err != nil {
+		log.Errorf("error tracking CIDs in hotstore: %s; falling back to coldstore", err)
+		return s.cold.PutMany(blks)
+	}
+
+	s.mx.Lock()
+	epoch := s.curTs.Height()
+	s.mx.Unlock()
+
+	batch := make([]cid.Cid, 0, len(blks))
+	for _, blk := range blks {
+		batch = append(batch, blk.Cid())
+	}
+
+	return s.snoop.PutBatch(batch, epoch)
+}
+
+func (s *SplitStore) AllKeysChan(ctx context.Context) (<-chan cid.Cid, error) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	chHot, err := s.hot.AllKeysChan(ctx)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	chCold, err := s.cold.AllKeysChan(ctx)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+
+	ch := make(chan cid.Cid)
+	go func() {
+		defer cancel()
+		defer close(ch)
+
+		for _, in := range []<-chan cid.Cid{chHot, chCold} {
+			for cid := range in {
+				select {
+				case ch <- cid:
+				case <-ctx.Done():
+					return
+				}
+			}
+		}
+	}()
+
+	return ch, nil
+}
+
+func (s *SplitStore) HashOnRead(enabled bool) {
+	s.hot.HashOnRead(enabled)
+	s.cold.HashOnRead(enabled)
+}
+
+func (s *SplitStore) View(cid cid.Cid, cb func([]byte) error) error {
+	err := s.hot.View(cid, cb)
+	switch err {
+	case bstore.ErrNotFound:
+		return s.cold.View(cid, cb)
+
+	default:
+		return err
+	}
+}
+
+// State tracking
+func (s *SplitStore) Start(cs *store.ChainStore) error {
+	s.cs = cs
+	s.curTs = cs.GetHeaviestTipSet()
+
+	// load base epoch from metadata ds
+	// if none, then use current epoch because it's a fresh start
+	bs, err := s.ds.Get(baseEpochKey)
+	switch err {
+	case nil:
+		s.baseEpoch = bytesToEpoch(bs)
+
+	case dstore.ErrNotFound:
+		err = s.setBaseEpoch(s.curTs.Height())
+		if err != nil {
+			return err
+		}
+
+	default:
+		return err
+	}
+
+	// watch the chain
+	cs.SubscribeHeadChanges(s.HeadChange)
+
+	return nil
+}
+
+func (s *SplitStore) Close() error {
+	if s.isCompacting() {
+		log.Warn("ongoing compaction; waiting for it to finish...")
+		for s.isCompacting() {
+			time.Sleep(time.Second)
+		}
+	}
+
+	return s.env.Close()
+}
+
+func (s *SplitStore) HeadChange(revert, apply []*types.TipSet) error {
+	s.mx.Lock()
+	s.curTs = apply[len(apply)-1]
+	epoch := s.curTs.Height()
+	s.mx.Unlock()
+
+	if !s.isCompacting() && epoch-s.baseEpoch > CompactionThreshold {
+		s.setCompacting(true)
```

Comment:
Can we please make `compacting` CAS? If there's a change upstream by which two HeadChanges arrive concurrently, we may launch two compactions.

Either that, or we guard the entire method in the mutex.

Reply:
I don't feel strongly about it, so we can make it CAS. It doesn't really matter though, it's 10 times per minute on average.